### ISSUE TITLE
jOOQ-meta Database implementations should dynamically detect server version and use appropriate dialect version

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9834]
The commercial jOOQ distributions have versioned dialects, which should be used by `jOOQ-meta`'s `Database` implementations to produce jOOQ SQL queries. This ensures that queries run by the code generator will work on older, supported database versions.

An example regression that was caused by the lack of this behaviour is https://github.com/jOOQ/jOOQ/issues/9832, where we stopped generating the `DUAL` table in MySQL queries starting from MySQL 8+, when in fact, the `MySQLDatabase` doesn't offer a simple way to configure the `SQLDialect.MYSQ_5_7` when used on older MySQL versions.

**Comment by lukaseder** [comment]
This is a catch-22 situation. The dialect needs to be known when creating the `DSLContext`. But the current methods of figuring out the database version is already using `DSLContext`, e.g. to detect `MYSQL_8`:

```java
    protected boolean is8() {

        // [#6602] The mysql.proc table got removed in MySQL 8.0
        if (is8 == null)
            is8 = !exists(PROC);

        return is8;
    }
```

Instead, we could query `JDBCUtils.dialect(Connection)`

**Comment by lukaseder** [comment]
Fixed in 3.14.0. Backport scheduled for 3.13.1 (#9837). I'm not scheduling a backport to 3.12.5, as this issue has never manifested before changing the `DUAL` behaviour, and patching the `SQLDialect` might just introduce new (rare) regressions, when users specify their own `SQLDialect` and `DSLContext` in custom subclasses of `AbstractDatabase`
